### PR TITLE
[Core]Cach named actor in local in case of getting them from GCS frequently

### PR DESCRIPTION
**Pull request by WangTaoTheTonic**
…uently

## Why are these changes needed?
Sometimes users would call `GetNamedActor` frequently, which will cause GCS pressure. To protect GCS we can cache the name locally.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/actor_manager.cc [hunk 1]
```diff
@@ -88,9 +89,12 @@ bool ActorManager::AddActorHandle(std::unique_ptr<ActorHandle> actor_handle,
         std::bind(&ActorManager::HandleActorStateNotification, this,
                   std::placeholders::_1, std::placeholders::_2);
     RAY_CHECK_OK(gcs_client_->Actors().AsyncSubscribe(
-        actor_id, actor_notification_callback, nullptr));
-  } else {
-    RAY_LOG(ERROR) << "Actor handle already exists " << actor_id.Hex();
```
Comment (MissiontoMars) [inline src/ray/core_worker/actor_manager.cc:93]: Should this log be kept?

File: src/ray/core_worker/actor_manager.h [hunk 2]
```diff
@@ -195,6 +203,9 @@ class ActorManager {
   /// Actor handle is a logical abstraction that holds actor handle's states.
   absl::flat_hash_map<ActorID, std::shared_ptr<ActorHandle>> actor_handles_
       GUARDED_BY(mutex_);
+
+  /// The map to cache the named actor in this worker locally..
```
Comment (MissiontoMars) [inline src/ray/core_worker/actor_manager.h:207]: typo

File: src/ray/core_worker/core_worker.cc [hunk 3]
```diff
@@ -2038,40 +2039,46 @@ std::pair<std::shared_ptr<const ActorHandle>, Status> CoreWorker::GetNamedActorH
...
+    RAY_CHECK_OK(gcs_client_->Actors().AsyncGetByName(
+        name, namespace_to_use,
+        [this, &actor_id, name, ready_promise](
+            Status status, const boost::optional<rpc::ActorTableData> &result) {
+          if (status.ok() && result) {
+            auto actor_handle = std::make_unique<ActorHandle>(*result);
+            actor_id = actor_handle->GetActorID();
+            actor_manager_->AddNewActorHandle(
+                std::move(actor_handle),
+                result.get().ray_namespace() + "-" + result.get().name(), GetCallerId(),
```
Comment (MissiontoMars) [inline src/ray/core_worker/core_worker.cc:2062]: Better to add comments about the rules of generating actor name.

**Comment by WangTaoTheTonic** [comment]
Failed tests seems unrelated.

File: src/ray/core_worker/actor_manager.cc (hunk 1 above)
Comment (WangTaoTheTonic) [reply src/ray/core_worker/actor_manager.cc:93]: We don't need this and it would confuse users.

File: src/ray/core_worker/actor_manager.cc [hunk 4]
```diff
@@ -168,4 +185,21 @@ std::vector<ObjectID> ActorManager::GetActorHandleIDsFromHandles() {
...
 
+ActorID ActorManager::GetCachedNamedActorID(const std::string &actor_name) {
+  {
+    absl::MutexLock lock(&cache_mutex_);
+    auto it = cached_actor_name_to_ids_.find(actor_name);
+    if (it != cached_actor_name_to_ids_.end()) {
+      absl::MutexLock lock(&mutex_);
+      auto handle_it = actor_handles_.find(it->second);
+      if (handle_it == actor_handles_.end()) {
+        cached_actor_name_to_ids_.erase(actor_name);
```
Comment (jovany-wang) [inline src/ray/core_worker/actor_manager.cc:196]: What's the case that the handle is in `cached_actor_name_to_ids_` but not in `actor_handles_`?

If we have no such case, maybe it's better to `CHECK()` to let the bug easy to figure.

**Review by jovany-wang — COMMENTED** [review]
On the high level structure, I prefer that core worker should be not aware of the cached_name, so that we have no necessary to be aware of the param `cached_name`.

In core worker, just get named actor from actor_manager.

**Comment by jovany-wang** [comment]
One more thing: let us add a basic case to get one actor twice? （I believe it's necessary to make sure the cache is work well.）

File: src/ray/core_worker/common.h [hunk 5]
```diff
@@ -32,6 +32,9 @@ std::string WorkerTypeString(WorkerType type);
 // Return a string representation of the language.
 std::string LanguageString(Language language);
 
+// Return a string representation of the named actor to cache.
+std::string ToCachedActorName(const std::string &ns, const std::string &actor_name);
```
[the remark this answers is not in the document]
Comment (WangTaoTheTonic) [reply src/ray/core_worker/common.h:36]: We already have a function `GetFullName` in jni_utils.h, which concat job id and actor id. Maybe `GenerateCachedActorName` is better here.

File: src/ray/core_worker/actor_manager.cc (hunk 4 above)
Comment (WangTaoTheTonic) [reply src/ray/core_worker/actor_manager.cc:196]: Alright. Such codes are used for sanity. CHECK() is better.

File: src/ray/core_worker/actor_manager.h [hunk 6]
```diff
@@ -167,7 +174,8 @@ class ActorManager {
   /// \param[in] actor_creation_return_id object id of this actor creation
   /// \return True if the handle was added and False if we already had a handle
   /// to the same actor.
-  bool AddActorHandle(std::unique_ptr<ActorHandle> actor_handle, bool is_owner_handle,
+  bool AddActorHandle(std::unique_ptr<ActorHandle> actor_handle,
+                      const std::string &cached_actor_name, bool is_owner_handle,
```
[the remark this answers is not in the document]
Comment (WangTaoTheTonic) [reply src/ray/core_worker/actor_manager.h:178]: Better to keep consistent with the function who generates this name?

**Review by jovany-wang — APPROVED** [review]
LGTM.

**Comment by jovany-wang** [comment]
All cases passed. There're 2 very minor comments which can be addressed in other PRs.

**Comment by rkooo567** [comment]
Sorry that I didn't notice this PR. Can you guys assign one of us (maybe me @wuisawesome or @iycheng) next time before merging changes in these layers? 

I have one question; why do we have a separate mutex for cache? `cache_mutex_ `? Can we just use a single mutex here? Were there some performance problems with that?

**Comment by jovany-wang** [comment]
> Sorry that I didn't notice this PR. Can you guys assign one of us (maybe me @wuisawesome or @iycheng) next time before merging changes in these layers?

Sorry for that we didn't inform you guys. Because I think this is a minor change, it's just cache something.
We'll notice you next time as you said.

And, I think it would great that if you guys could inform us for some related changes as well. Our guys are also actively to contribute on these layer as well.

> I have one question; why do we have a separate mutex for cache? `cache_mutex_ `? Can we just use the original mutex here? Were there some performance problems with that?

@WangTaoTheTonic

**Comment by WangTaoTheTonic** [comment]
@rkooo567 They protect different objects so we use separate mutex. It's uncommon that use one same mutex to lock different objects unless they are always accessed together.

**Comment by rkooo567** [comment]
I see. I did realize they are not accessed together with the main actor manager mutex. Thanks for the explanation.

**Comment by rkooo567** [comment]
Hmm actually I saw this code path;

```cpp
    absl::MutexLock lock(&cache_mutex_);
    auto it = cached_actor_name_to_ids_.find(actor_name);
    if (it != cached_actor_name_to_ids_.end()) {
      absl::MutexLock lock(&mutex_);
      auto handle_it = actor_handles_.find(it->second);
      RAY_CHECK(handle_it != actor_handles_.end());
      return it->second;
    }
  }
```

For the current code, it is fine, but it is really easy to introduce deadlock when new code is added (if there's logic that acquires mutex first and cache_mutex after). Is there any performance concern of using one mutex here? Otherwise, I prefer to just use &mutex to protect this field.

**Pull request merged**